src/adapters/items-fetcher.ts

```diff
@@ -0,0 +1,158 @@
+import LRU from 'lru-cache'
+import { AppComponents, Limits, ItemsResult, Item, ItemFetcher, ItemFetcherError, ItemFetcherErrorCode } from '../types'
+import { ISubgraphComponent } from '@well-known-components/thegraph-component'
+import { compareByRarity } from '../logic/utils'
+
+const THE_GRAPH_PAGE_SIZE = 1000
+
+// TODO cache metrics
+type ItemsQueryResponse = {
+  nfts: ItemFromQuery[]
+}
+
+type ItemFromQuery = {
+  urn: string
+  id: string
+  tokenId: string
+  transferredAt: number
+  item: {
+    rarity: string
+    price: number
+  }
+}
+
+const QUERY_EMOTES = createQueryForCategory('emote')
+const QUERY_WEARABLES = createQueryForCategory('wearable')
+
+function groupItemsByURN(items: ItemFromQuery[]): Item[] {
+  const itemsByURN = new Map<string, Item>()
+
+  items.forEach((item) => {
+    const individualData = {
+      id: item.id,
+      tokenId: item.tokenId,
+      transferredAt: item.transferredAt,
+      price: item.item.price
+    }
+    if (itemsByURN.has(item.urn)) {
+      const itemFromMap = itemsByURN.get(item.urn)!
+      itemFromMap.individualData.push(individualData)
+      itemFromMap.amount = itemFromMap.amount + 1
+    } else {
+      itemsByURN.set(item.urn, {
+        urn: item.urn,
+        individualData: [individualData],
+        rarity: item.item.rarity,
+        amount: 1
+      })
+    }
+  })
+
+  return Array.from(itemsByURN.values())
+}
+
+type ItemCategory = 'wearable' | 'emote'
+
+function createQueryForCategory(category: ItemCategory) {
+  return `query fetchItemsByOwner($owner: String, $idFrom: String) {
+    nfts(
+      where: { id_gt: $idFrom, owner: $owner, category: "${category}"},
+      orderBy: id,
+      orderDirection: asc,
+      first: ${THE_GRAPH_PAGE_SIZE}
+    ) {
+      urn,
+      id,
+      tokenId,
+      category
+      transferredAt,
+      item {
+        rarity,
+        price
+      }
+    }
+  }`
+}
+
+async function runItemsQuery(
+  subgraph: ISubgraphComponent,
+  address: string,
+  category: ItemCategory
+): Promise<ItemFromQuery[]> {
+  const items = []
+  const owner = address.toLowerCase()
+  let idFrom = ''
+  let result: ItemsQueryResponse
+  const query = category === 'wearable' ? QUERY_WEARABLES : QUERY_EMOTES
+  do {
+    result = await subgraph.query<ItemsQueryResponse>(query, {
+      owner,
+      idFrom
+    })
+
+    if (result.nfts.length === 0) {
+      break
+    }
+
+    for (const nft of result.nfts) {
+      items.push(nft)
+    }
+
+    idFrom = items[items.length - 1].id
+  } while (result.nfts.length === THE_GRAPH_PAGE_SIZE)
```

Comment:
Another refactor option to remove the conditional within the do:

```
  do {
    result = await subgraph.query<ItemsQueryResponse>(query, {
      owner,
      idFrom
    })

    for (const nft of result.nfts) {
      items.push(nft)
    }

    idFrom = items[items.length - 1].id
  } while (result.nfts.length === THE_GRAPH_PAGE_SIZE || result.nfts.length !== 0)
```

Reply:
I think this would file in the items[items.lenght - 1] if the query returned no results

Reply:
i don't understand @hugoArregui. "file" or "fail"?

Reply:
`fail` sorry, I was talking about Ale's comment though